### PR TITLE
WhatsApp notifier class added, comnfiguration form started

### PR DESCRIPTION
**Pull request by stefanodellosa-personal**
Introduce WhatsApp notification type to WADAS

### CODE REVIEW

File: wadas/domain/configuration.py [hunk 1]
```diff
@@ -59,7 +60,27 @@ def load_configuration_from_file(file_path):
                     email_notifier.sender_email,
                 )
                 valid_email_keyring = False
-
+        elif key in Notifier.notifiers and key == Notifier.NotifierTypes.WHATSAPP.value:
+            whatsapp_notifier = WhatsAppNotifier(**value)
+            Notifier.notifiers[key] = whatsapp_notifier
+            credentials = keyring.get_credential("WADAS_WhatsApp", whatsapp_notifier.sender_id)
+            if not credentials:
```
Comment (cdimauro) [inline wadas/domain/configuration.py:67]: You can combine them with the := operator.

File: wadas/domain/configuration.py [hunk 2]
```diff
@@ -59,7 +60,27 @@ def load_configuration_from_file(file_path):
...
                 )
                 valid_email_keyring = False
-
+        elif key in Notifier.notifiers and key == Notifier.NotifierTypes.WHATSAPP.value:
+            whatsapp_notifier = WhatsAppNotifier(**value)
+            Notifier.notifiers[key] = whatsapp_notifier
+            credentials = keyring.get_credential("WADAS_WhatsApp", whatsapp_notifier.sender_id)
+            if not credentials:
+                logger.error(
+                    "Unable to find WhatsApp credentials for %s stored on the system."
```
Comment (cdimauro) [inline wadas/domain/configuration.py:69]: Add a space after the .

File: wadas/domain/email_notifier.py [hunk 3]
```diff
@@ -83,6 +83,20 @@ def send_email(self, img_path):
...
+
+    def is_configured(self):
+        """Method that returns configuration status as bool value."""
+
+        username = self.sender_email
+        credentials = keyring.get_credential("WADAS_email", username)
+        if self.smtp_hostname and self.smtp_port and self.recipients_email and credentials.username:
+            return True
+        else:
+            return False
```
Comment (cdimauro) [inline wadas/domain/email_notifier.py:98]: return bool(self.smtp_hostname and self.smtp_port and self.recipients_email and credentials.username)

File: wadas/domain/whatsapp_notifier.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            return
+
+        headers = {
+            "Authorization": f"Bearer {credentials.password}",
+            "Content-Type": "application/json",
+        }
+
+        media_id = None
+        if self.allow_images:
+            media_id = self.load_image(credentials.password, img_path)
```
Comment (cdimauro) [inline wadas/domain/whatsapp_notifier.py:60]: media_id = self.load_image(credentials.password, img_path) if self.allow_images else None

File: wadas/ui/configure_whatsapp_dialog.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
+"""Whatsapp configuration module."""
+
+
+import os
+import requests
```
Comment (cdimauro) [inline wadas/ui/configure_whatsapp_dialog.py:5]: Insert blank line to separate imports.

File: wadas/ui/configure_whatsapp_dialog.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+                self.ui.label_errorMessage.setText("Recipients number list cannot be empty.")
+            self.ui.buttonBox.button(QDialogButtonBox.Ok).setEnabled(False)
+            self.ui.pushButton_testMessage.setEnabled(False)
+
+    def add_whatsapp_credentials(self, username, token):
+        """Method to add WhatsApp credentials to system keyring."""
+
+        # If credentials exist remove them (workaround keyring bug)
+        credentials = keyring.get_credential(
+            f"WADAS_WhatsApp", ""
```
Comment (cdimauro) [inline wadas/ui/configure_whatsapp_dialog.py:81]: f"WADAS_WhatsApp" -> "WADAS_WhatsApp"

File: wadas/ui/configure_whatsapp_dialog.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        )
+        if credentials:
+            try:
+                keyring.delete_password(f"WADAS_WhatsApp", credentials.username)
```
Comment (cdimauro) [inline wadas/ui/configure_whatsapp_dialog.py:85]: Same here.

File: wadas/ui/configure_whatsapp_dialog.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+            "WADAS_WhatsApp",
+            username,
+            token,
+        )
+
+    def accept_and_close(self):
+        """When Ok is clicked, save email config info before closing."""
+        recipients = []
+        for recipient in self.ui.textEdit_recipientsNumbers.toPlainText().strip().split(", "):
+            recipients.append(recipient)
```
Comment (cdimauro) [inline wadas/ui/configure_whatsapp_dialog.py:101]: recipients = self.ui.textEdit_recipientsNumbers.toPlainText().strip().split(", ")

File: wadas/ui/configure_whatsapp_dialog.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+    def send_whatsapp_message(self):
+        """Method to send WhatsApp message notification."""
+
+        message = "WADAS Test Message!"
+        url = f"https://graph.facebook.com/v17.0/{self.ui.lineEdit_phoneID.text()}/messages"
+        access_token = self.ui.lineEdit_accessToken.text()
+        recipients = []
+        for recipient in self.ui.textEdit_recipientsNumbers.toPlainText().strip().split(", "):
+            recipients.append(recipient)
```
Comment (cdimauro) [inline wadas/ui/configure_whatsapp_dialog.py:130]: recipients = self.ui.textEdit_recipientsNumbers.toPlainText().strip().split(", ")

**Review by cdimauro — APPROVED** [review]

**Review by hexfati — APPROVED** [review]
LGTM but I would probably write a line in the log when sending the notification (WhatsApp or email) to keep track on the GUI.

**Comment by stefanodellosa-personal** [comment]
> LGTM but I would probably write a line in the log when sending the notification (WhatsApp or email) to keep track on the GUI.

It is logged in wadas/domain/whatsapp_notifier.py line 88

**Pull request merged**